test/e2e/framework.go

```diff
@@ -949,12 +949,13 @@ func (data *TestData) waitForAntreaDaemonSetPods(timeout time.Duration) error {
 		}
 
 		// Make sure that all Daemon Pods are available.
-		// We use clusterInfo.numNodes instead of DesiredNumberScheduled because
-		// DesiredNumberScheduled may not be updated right away. If it is still set to 0 the
-		// first time we get the DaemonSet's Status, we would return immediately instead of
-		// waiting.
-		desiredNumber := int32(clusterInfo.numNodes)
-		if currentNumAvailable != desiredNumber || UpdatedNumberScheduled != desiredNumber {
+		if dsLinux != nil && dsLinux.Status.ObservedGeneration != dsLinux.Generation {
+			return false, nil
+		}
+		if dsWindows != nil && dsWindows.Status.ObservedGeneration != dsWindows.Generation {
+			return false, nil
+		}
+		if currentNumAvailable != UpdatedNumberScheduled {
```

Comment:
These numbers could be out-dated when we delete the agent Pods directly. That's why we used number of Nodes as the desired number. Now that we don't know the number of Nodes, I meant we should switch back to use the desired number in the DaemonSet status, but add a check for the generation to fix the previous problem (when the generation matches, the desired number is reliable). So we don't need to check the other numbers at all.

```
var desiredNum int32
if dsLinux, err = getDS(antreaDaemonSet, "Linux"); err != nil {
	return false, err
}
if dsLinux.Generation != dsLinux.Status.ObservedGeneration {
	return false, nil
}
desiredNum += dsLinux.Status.DesiredNumberScheduled

if len(clusterInfo.windowsNodes) != 0 {
	var dsWindows *appsv1.DaemonSet
	if dsWindows, err = getDS(antreaWindowsDaemonSet, "Windows"); err != nil {
		return false, err
	}
	if dsWindows.Generation != dsWindows.Status.ObservedGeneration {
		return false, nil
	}
	desiredNum += dsWindows.Status.DesiredNumberScheduled
}

// Make sure that all antrea-agent Pods are not terminating. This is required because NumberAvailable of
// DaemonSet counts Pods even if they are terminating. Deleting antrea-agent Pods directly does not cause the
// number to decrease if the process doesn't quit immediately, e.g. when the signal is caught by bincover
// program and triggers coverage calculation.
pods, err := data.clientset.CoreV1().Pods(antreaNamespace).List(context.TODO(), metav1.ListOptions{
	LabelSelector: "app=antrea,component=antrea-agent",
})
if err != nil {
	return false, fmt.Errorf("failed to list antrea-agent Pods: %v", err)
}
if len(pods.Items) != int(desiredNum) {
	return false, nil
}
for _, pod := range pods.Items {
	if pod.DeletionTimestamp != nil {
		return false, nil
	}
}
return true, nil
```